### PR TITLE
Develop complex signals

### LINKED ISSUE #384: Roadmap to complex signals
Problem: We want to allow signals with complex time data 

Required changes in audio classes:
- [x] TimeData: add `complex` argument in init and make it a class property
- [x] Signal.domain: check `complex` and chose FFT accordingly                           
- [x] Signal.freq: Call normalization according to `complex` property
- [x] FFT module: wrap fft and ifft and fft_frequencies in analogy to the real parts
- [x] Frequencies in complex signals will go from -fs/2 to fs/2
- [x] check arithmetic operations on audio class, with one complex and one real-valued signal.

Changes in plot module (good first issues)
- [x] Plot functions time domain: add parameter `???` to plot `'real'`, `'imag'`, or `'abs'` values
- [x] Plot functions frequency domain: add parameter `side` to plot `'left'` or `'right'` sided spectrum
- [x] Add plot interaction for both parameters

Add tests in dsp module (good first issues)
- [x] average
- [x] convolve
- [x] decibel
- [x] deconvolve
- [x] fractional_time_shift
- [x] phase
- [x] pad_zeros
- [x] regularized_spectrum_inversion
- [x] resample (depends on scipy implementation)
- [x] spectrogram
- [x] time_shift
- [x] time_window
- [x] filter classes

Throw errors in dsp module - we decided to implement energy, power, and rms for complex signals
- [x] energy
- [x] power
- [x] rms
 
Possibily change functionality
- [x] find_impulse_response_delay:
- [x] find_impulse_response_start: detect separately for real and imag and return earliest or use absolute value (test with gammatone filters and decide later, probably absolute value because this is the envelope)
- [x] group_delay (check because the estimation is FFT based)
- [x] normalize (some modes like rms not supported)
- [x] smooth_fractional_octaves
- [x] update concatenate function (see comment in #452)

Open question (throw errors for now)
- [x] linear_phase
- [x] minimum_phase
- [x] zero_phase

To discuss:
- arithmetic operation that casts a real to complex valued signal, with invalid ftt_norm. Cast to none or error?

To do:
- check docstrings for all above methods

### PR DESCRIPTION
**Pull request by f-brinkmann**
### Which issue(s) are closed by this pull request?

Closes #384 

### Changes proposed in this pull request:

This is a branch for collecting approved pull requests related to #384. It will be merged once all the tasks are done

### CODE REVIEW

**Comment by ahms5** [comment]
merged dev into this branch, #704 required to fix ruff

File: pyfar/dsp/dsp.py [hunk 1]
```diff
@@ -304,10 +304,25 @@ def spectrogram(signal, window='hann', window_length=1024,
     window = sgn.get_window(window, window_length)
     hop = window_length - window_overlap
 
-    frequencies, times, spectrogram = sgn.spectrogram(
-        x=signal.time, fs=signal.sampling_rate, window=window,
-        noverlap=window_overlap, mode='magnitude', scaling='spectrum',
-        return_onesided=not signal.complex)
+    fft_mode = 'onesided'
+    if signal.complex:
+        fft_mode = 'twosided'
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:309]: `fft_mode = 'twosided' if signal.complex else 'onesided'`

**Review by f-brinkmann — COMMENTED** [review]
Thanks - found only one line for improvement. Everything fine apart from that.

**Review by tluebeck — DISMISSED** [review]

File: pyfar/classes/audio.py [hunk 2]
```diff
@@ -655,10 +712,13 @@ class Signal(FrequencyData, TimeData):
         or ``'psd'``. See :py:func:`~pyfar.dsp.fft.normalization` and [#]_
         for more information. The default is ``'none'``, which is typically
         used for energy signals, such as impulse responses.
-    comment : str
+    comment : str, optional
         A comment related to `data`. The default is ``""``, which
         initializes an empty string.
-
+    is_complex : bool
```
Comment (ahms5) [inline pyfar/classes/audio.py:718]:
```suggestion
    is_complex : bool, optional
```

**Review by ahms5 — APPROVED** [review]
Thanks, I just found one small thing

**Review by ahms5 — APPROVED** [review]

**Pull request merged**